src/Features/Core/Portable/SemanticSearch/Tools/ReferencingSyntaxFinder.cs

```diff
@@ -34,85 +32,39 @@ public IEnumerable<SyntaxNode> Find(ISymbol symbol)
 
     public async IAsyncEnumerable<SyntaxNode> FindAsync(ISymbol symbol)
     {
-        var channel = Channel.CreateUnbounded<ReferenceLocation>(new()
+        using var _ = PooledHashSet<SyntaxNode>.GetInstance(out var cachedRoots);
+
+        // Kick off the SymbolFinder.FindReferencesAsync call on the provided symbol/solution.  As it finds
+        // ReferenceLocations, it will push those into the 'callback' delegate passed into it. ProducerConsumer will
+        // then convert this to a simple IAsyncEnumerable<ReferenceLocation> that we can iterate over, converting those
+        // locations to SyntaxNodes in the corresponding C# or VB document.
+        await foreach (var item in ProducerConsumer<ReferenceLocation>.RunAsync(
+            FuncReferencesAsync, args: (solution, symbol), cancellationToken))
         {
-            SingleReader = true,
-            SingleWriter = false,
-        });
-
-        using var _ = cancellationToken.Register(
-            static (obj, cancellationToken) => ((Channel<SourceReferenceItem>)obj!).Writer.TryComplete(new OperationCanceledException(cancellationToken)),
-            state: channel);
```

Comment:
ProducerConsumer.RunChannelAsync basically has all this code, and is the centralized place where so much of our Producer/Consumer logic goes through.  I added a new entrypoint to it to represent this case (where caller just wants to grab the produced results and do transformations on it itself).